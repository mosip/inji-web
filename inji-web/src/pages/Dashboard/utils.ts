export const convertStringIntoPascalCase = (text: string | null) => {
    if (text != null) {
        return text
            .toLocaleLowerCase()
            .split(" ")
            .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
            .join(" ");
    }
};
