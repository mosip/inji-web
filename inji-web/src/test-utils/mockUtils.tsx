import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, BrowserRouter as Router, Routes } from 'react-router-dom';
import { reduxStore } from '../redux/reduxStore';

// Mock for storage module
export const mockStorageModule = () => {
    jest.mock('../utils/storage.ts', () => ({
        storage: {
            getItem: jest.fn(),
            setItem: jest.fn(),
            SESSION_INFO: 'SESSION_INFO',
            SELECTED_LANGUAGE: 'selectedLanguage'
        }
    }));
};

// Mock for localStorage
export const mockLocalStorage = () => {
    let store: { [key: string]: string } = {};
  
    const localStorageMock = {
      getItem: jest.fn((key: string) => store[key] || null),
      setItem: jest.fn((key: string, value: string) => {
        store[key] = value;
      })
    };
  
    Object.defineProperty(window, 'localStorage', { 
      value: localStorageMock,
      writable: true
    });
  
    return localStorageMock;
  };

// API Mock
export const mockApi = {
    authorizationRedirectionUrl: 'http://mockurl.com'
};

// Crypto Mock
export const mockCrypto = {
    getRandomValues: (array: Uint32Array) => {
        for (let i = 0; i < array.length; i++) {
            array[i] = 4294967296;
        }
        return array;
    },
    subtle: {} as SubtleCrypto,
    randomUUID: () => '123e4567-e89b-12d3-a456-426614174000'
} as Crypto;

export const mockUseTranslation = () => {
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
            t: (key: string) => key,
        }),
    }));
};

export const mockUseNavigate = () => {
    const mockNavigate = jest.fn();
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useNavigate: mockNavigate,
    }));
};

export const mockUseSelector = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useSelector: jest.fn(),
    }));
};

export const setMockUseSelectorState = (state: any) => {
    const useSelectorMock = require('react-redux').useSelector;
    useSelectorMock.mockImplementation((selector: any) => selector(state));
};

export const mockUseGetObjectForCurrentLanguage = () => {
    jest.mock('../utils/i18n', () => ({
        getObjectForCurrentLanguage: jest.fn(),
    }));
};

export const wrapUnderRouter = (children: React.ReactNode) => {
    return <Router>{children}</Router>;
};

export const mockUseSearchCredentials = () => {
    jest.mock('../components/Credentials/SearchCredential', () => ({
        SearchCredential: () => <></>,
    }));
 };

export const mockUseDispatch = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useDispatch: jest.fn(),
    }));
};
              
export const mockUseSpinningLoader = () => {
    jest.mock('../components/Common/SpinningLoader', () => ({
        SpinningLoader: () => <></>,
    }));
};

export const mockUseLanguageSelector = () => {
    jest.mock('../components/Common/LanguageSelector', () => ({
        LanguageSelector: () => <></>,
    }));
};

interface RenderWithProviderOptions extends Omit<RenderOptions, 'queries'> {}

export const renderWithProvider = (element: ReactElement, options?: RenderWithProviderOptions) => {
    return render(
        <Provider store={reduxStore}>
            <Router>
                {element}
            </Router>
        </Provider>,
        options
    );
};
    
export const mockUseParams = ()=>{
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useParams: () => ({ issuerId: 'test-issuer-id' }),
      }));   
};

export const mockUseapiObject = () =>{
    jest.mock('../utils/api', () => ({
        api: {
          mimotoHost: 'http://mocked-api-host',
        },
      }));
};

export const mockUseFetch = () =>{
    jest.mock('../hooks/useFetch', () => {
        const RequestStatus = {
          LOADING: 'LOADING',
          DONE: 'DONE',
          ERROR: 'ERROR',
        };
        return {
          useFetch: () => jest.fn(),
          RequestStatus,
        };
      });
};

export const mockUseToast = () =>{
    jest.mock('react-toastify', () => ({
      toast: {
        error: jest.fn(),
      },
    }));
};

export const mockusemisc = ()=>{
    jest.mock('../utils/misc', () => ({
        downloadCredentialPDF: jest.fn(),
        getErrorObject: jest.fn(),
        getTokenRequestBody: jest.fn(),
      }));
};

export const mockusei18n = ()=>{
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
          t: (key: string) => key,
        }),
        initReactI18next: {
          type: '3rdParty',
          init: jest.fn(),
        },
      }));

};

export const mockWindowLocation = (url: string) => {
    const location = new URL(url) as unknown as Location;
    Object.defineProperty(window, 'location', {
      value: location,
      writable: true,
    });
  };

export const renderWithRouter = (Element: React.ReactElement, { route = '/' } = {}) => {
    window.history.pushState({}, 'Test page', route);
    return render(
      <BrowserRouter>
        <Provider store={reduxStore}>
          <Routes>
            <Route path="*" element={Element} />
          </Routes>
        </Provider>
      </BrowserRouter>
    );
  };

