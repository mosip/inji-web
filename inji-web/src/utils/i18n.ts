import i18n from "i18next";
import {initReactI18next} from "react-i18next";
import en from '../locales/en.json';
import fr from '../locales/fr.json';
import ta from '../locales/ta.json';
import hi from '../locales/hi.json';
import kn from '../locales/kn.json';
import ar from '../locales/ar.json';
import {storage} from "./storage";
import {DisplayArrayObject, LanguageObject} from "../types/data";

const resources = {en, ta, kn, hi, fr, ar};

export const LanguagesSupported: LanguageObject[] = [
    {label: "English", value: 'en'},
    {label: "தமிழ்", value: 'ta'},
    {label: "ಕನ್ನಡ", value: 'kn'},
    {label: "हिंदी", value: 'hi'},
    {label: "Français", value: 'fr'},
    {label: "عربي", value: 'ar'}
]

export const defaultLanguage = window._env_.DEFAULT_LANG;

const selected_language = storage.getItem(storage.SELECTED_LANGUAGE);
i18n
    .use(initReactI18next) // passes i18n down to react-i18next
    .init({
        resources,
        lng: selected_language ? selected_language : defaultLanguage,
        fallbackLng: defaultLanguage,
        interpolation: {
            escapeValue: false // react already safes from xss => https://www.i18next.com/translation-function/interpolation#unescape
        },
    });

export const switchLanguage = async (language: string) => {
    storage.setItem(storage.SELECTED_LANGUAGE, language);
    await i18n.changeLanguage(language);
}
export const getObjectForCurrentLanguage = (displayArray: DisplayArrayObject[], language: string = i18n.language) => {
    let resp = displayArray.filter(displayObj => (displayObj.language === language || displayObj.locale === language))[0];
    if (!resp) {
        resp = displayArray.filter(displayObj => (displayObj.language === defaultLanguage || displayObj.locale === defaultLanguage))[0];
    }
    return resp;
}

export const getDirCurrentLanguage = (language: string) => {
    return isRTL(language) ? 'rtl' : 'ltr';
}

export const isRTL = (language:string) => {
    return language === 'ar';
}
