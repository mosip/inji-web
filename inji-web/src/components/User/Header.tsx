import React, {useEffect, useRef, useState} from 'react';
import {useNavigate} from 'react-router-dom';
import {LanguageSelector} from '../Common/LanguageSelector';
import {api} from '../../utils/api';
import {toast} from 'react-toastify';
import {useUser} from '../../hooks/User/useUser';
import HamburgerMenu from '../../assets/HamburgerMenu.svg';
import {isRTL} from '../../utils/i18n';
import {useSelector} from 'react-redux';
import {RootState} from '../../types/redux';
import {useTranslation} from 'react-i18next';
import DropdownArrowIcon from '../Common/DropdownArrowIcon';
import {ROUTES} from '../../utils/constants';
import {convertStringIntoPascalCase} from "../../utils/misc";
import {navigateToUserHome} from "../../utils/navigationUtils";
import {CircleSkeleton} from '../Common/CircleSkeleton';
import {InfoFieldSkeleton} from '../Common/InfoFieldSkeleton';
import {ApiError, DropdownItem, ResponseTypeObject} from "../../types/data";
import {useApi} from "../../hooks/useApi";

type UserHeaderProps = {
    headerRef: React.RefObject<HTMLDivElement>;
    headerHeight: number;
};

export const Header: React.FC<UserHeaderProps> = ({
    headerRef,
    headerHeight
}) => {
    const language = useSelector((state: RootState) => state.common.language);
    const navigate = useNavigate();
    const [displayName, setDisplayName] = useState<string | undefined>(
        undefined
    );
    const [isProfileDropdownOpen, setIsProfileDropdownOpen] = useState(false);
    const [isHamburgerMenuOpen, setIsHamburgerMenuOpen] = useState(false);
    const dropdownRef = useRef<HTMLDivElement>(null);
    const hamburgerMenuRef = useRef<HTMLImageElement>(null);
    const {user, removeUser,isLoading} = useUser();
    const displayNameFromLocalStorage = user?.displayName;
    const hasProfilePictureUrl = user?.profilePictureUrl;
    const {t} = useTranslation('User');
    const logoutRequest = useApi()

    useEffect(() => {
        setDisplayName(displayNameFromLocalStorage);
    }, [displayNameFromLocalStorage]);

    useEffect(() => {
        const handleClickOutside = (event: MouseEvent) => {
            const target = event.target as Node;

            if (dropdownRef.current && !dropdownRef.current.contains(target)) {
                setIsProfileDropdownOpen(false);
            }

            if (
                hamburgerMenuRef.current &&
                !hamburgerMenuRef.current.contains(target)
            ) {
                setIsHamburgerMenuOpen(false);
            }
        };

        document.addEventListener('click', handleClickOutside);
        return () =>
            document.removeEventListener('click', handleClickOutside);
    }, []);

    const handleLogout = async () => {
        try {
            const response = await logoutRequest.fetchData({
                apiRequest: api.userLogout,
            })

            if (response.ok()) {
                removeUser();
                window.location.replace(ROUTES.ROOT);
            } else {
                const parsedResponse = ((response.error as ApiError)?.response?.data as ResponseTypeObject).errors
                const errorCode = parsedResponse?.[0].errorCode;
                if (errorCode === 'user_logout_error') {
                    removeUser();
                    window.location.replace(ROUTES.ROOT);
                }
                throw new Error(parsedResponse?.[0]?.errorMessage);
            }
        } catch (error) {
            console.error('Logout failed with error:', error);
            toast.error('Unable to logout. Please try again.');
        }
    };

    const dropdownItems: DropdownItem[] = [
        {
            label: t('ProfileDropdown.profile'),
            onClick: () => {
                setIsProfileDropdownOpen(false);    
                navigate(ROUTES.PROFILE,{state: {from: window.location.pathname}});
            },
            textColor: 'text-gray-700',
            key: 'Profile-Dropdown-Profile'
        },
        {
            label: t('ProfileDropdown.faq'),
            onClick: () => {
                setIsProfileDropdownOpen(false);
                navigate(ROUTES.USER_FAQ, {state: {from: window.location.pathname}});
            },
            textColor: 'text-gray-700',
            key: 'Profile-Dropdown-FAQ'
        },
        {
            label: t('ProfileDropdown.logout'),
            onClick: handleLogout,
            textColor: 'text-red-700',
            key: 'Profile-Dropdown-Logout'
        }
    ];

    const toggleProfileDropdown = (e: React.MouseEvent<HTMLDivElement>) => {
        e.stopPropagation(); // Prevent global click handler from immediately closing the dropdown
        setIsProfileDropdownOpen(!isProfileDropdownOpen);
    };

    const toggleHamburgerMenu = () => {
        setIsHamburgerMenuOpen(!isHamburgerMenuOpen);
    };

    const getProfileInitials = (displayName: string | undefined) => {
        return displayName ? displayName.charAt(0).toUpperCase(): 'U';
    };

    const getUserProfileIconWithName = () => {
        if (isLoading) {
            return (
              <div className="flex gap-2 items-center">
                <CircleSkeleton size="w-12 h-12" />
                <InfoFieldSkeleton width="w-24" height="h-2" />
              </div>
            );
          }
        
          return (
            <div className="flex gap-2 items-center">
              <div
                className={`aspect-square w-12 rounded-full bg-iw-avatarPlaceholder overflow-hidden flex items-center justify-center text-iw-avatarText font-medium text-lg ${
                  !hasProfilePictureUrl && 'p-2 sm:p-3 md:p-4'
                }`}
              >
                {hasProfilePictureUrl ? (
                  <img
                    src={user.profilePictureUrl}
                    alt="Profile Pic"
                    className="w-12 h-12 object-cover rounded-full"
                    referrerPolicy="no-referrer"
                  />
                ) : (
                  getProfileInitials(displayName)
                )}
              </div>
              <span className="font-semibold text-gray-800">
                {convertStringIntoPascalCase(displayName)}
              </span>
            </div>
          );
        };

    return (
        <header
            ref={headerRef}
            data-testid="dashboard-header-container"
            className="fixed top-0 left-0 right-0 z-20 bg-iw-background bg-transparent shadow-[0_4px_5px_0_rgba(0,0,0,0.051)]"
        >
            <div className="container mx-auto px-4 py-4 flex justify-between items-center">
                <div className="flex justify-start items-center gap-2">
                    <div
                        data-testid="hamburger-menu"
                        className="block sm:hidden"
                    >
                        <img
                            data-testid="icon-hamburger-menu"
                            ref={hamburgerMenuRef}
                            src={HamburgerMenu}
                            alt="Hamburger Menu"
                            onClick={toggleHamburgerMenu}
                            className="cursor-pointer"
                        />
                    </div>
                    <div
                        className="w-[130px] sm:w-[160px] md:w-[170px]"
                        role={'button'}
                        tabIndex={0}
                        onMouseDown={() => navigateToUserHome(navigate)}
                        onKeyUp={() => navigateToUserHome(navigate)}
                    >
                        <img
                            src={require('../../assets/InjiWebLogo.png')}
                            className={`max-w-full h-auto object-contain cursor-pointer ${
                                isRTL(language) ? 'mr-4' : ''
                            }`}
                            data-testid="header-injiWeb-logo"
                            alt="Inji Web Logo"
                        />
                    </div>
                </div>

                <div className="flex items-center space-x-6">
                    <LanguageSelector />

                    <div
                        data-testid="profile-details"
                        className="hidden sm:block relative"
                        ref={dropdownRef}
                    >
                        <div className="flex items-center space-x-2">
                            {getUserProfileIconWithName()}
                            {!isLoading && (
                                <div
                                className="relative inline-block cursor-pointer"
                                onClick={toggleProfileDropdown}
                                >
                                <DropdownArrowIcon isOpen={isProfileDropdownOpen} />
                                </div>
                            )}
                        </div>

                        {isProfileDropdownOpen && (
                            <div
                                data-testid="profile-dropdown"
                                className="absolute -right-7 top-100 mt-8 w-56 bg-white rounded-lg shadow-lg z-50 font-medium"
                            >
                                <div className="absolute top-[-0.3rem] right-8 w-3 h-3 bg-white transform rotate-45" />
                                <div className="py-2">
                                    {dropdownItems.map((item, index) => (
                                        <React.Fragment key={item.key}>
                                            <div
                                                className={`px-4 py-2 text-sm ${item.textColor} hover:bg-gray-100 cursor-pointer`}
                                                onClick={item.onClick}
                                            >
                                                {item.label}
                                            </div>
                                            {index !==
                                                dropdownItems.length - 1 && (
                                                <hr className="border-gray-200 my-1 mx-2" />
                                            )}
                                        </React.Fragment>
                                    ))}
                                </div>
                            </div>
                        )}
                    </div>
                </div>
            </div>
            <div
                data-testid="hamburger-menu-dropdown"
                className="block sm:hidden w-full"
            >
                {isHamburgerMenuOpen && (
                    <div
                        style={{marginTop: headerHeight}}
                        className="absolute top-1 bg-white shadow-iw-hamburger-dropdown p-2 w-full"
                    >
                        <div>
                            <div className="flex items-center px-4 py-2">
                                {getUserProfileIconWithName()}
                            </div>
                            {dropdownItems.map((item, index) => (
                                <React.Fragment key={item.key}>
                                    <div
                                        className={`px-4 py-2 text-sm ${item.textColor} hover:bg-gray-100 cursor-pointer font-medium`}
                                        onClick={item.onClick}
                                    >
                                        {item.label}
                                    </div>
                                    {index !== dropdownItems.length - 1 && (
                                        <hr className="border-gray-200 my-1 mx-2" />
                                    )}
                                </React.Fragment>
                            ))}
                        </div>
                    </div>
                )}
            </div>
        </header>
    );
};
