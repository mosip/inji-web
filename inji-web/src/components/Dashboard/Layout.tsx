import React, {useEffect, useRef, useState} from 'react';
import {Sidebar} from './Sidebar';
import {Header} from './Header';
import {Footer} from '../PageTemplate/Footer';
import {useSelector} from 'react-redux';
import {RootState} from '../../types/redux';
import {getDirCurrentLanguage} from '../../utils/i18n';
import {Outlet} from 'react-router-dom';
import DashboardBgTop from '../../assets/Background.svg';
import DashboardBgBottom from '../../assets/DashboardBgBottom.svg';

export const Layout: React.FC = () => {
    const language = useSelector((state: RootState) => state.common.language);
    const headerRef = useRef<HTMLDivElement>(null);
    const footerRef = useRef<HTMLDivElement>(null);
    const [headerHeight, setHeaderHeight] = useState(0);
    const [footerHeight, setFooterHeight] = useState(0);

    useEffect(() => {
        if (headerRef.current) {
            setHeaderHeight(headerRef.current.getBoundingClientRect().height);
        }
        if (footerRef.current) {
            setFooterHeight(footerRef.current.getBoundingClientRect().height);
        }
    }, []);

    return (
        <div
            className="h-screen flex flex-col bg-iw-background font-base overflow-hidden w-full relative"
            dir={getDirCurrentLanguage(language)}
        >
            <Header ref={headerRef} headerHeight={headerHeight} />

            <div
                className="flex flex-1 overflow-hidden w-full relative"
                style={{
                    marginTop: headerHeight,
                    marginBottom: footerHeight,
                    zIndex: 0
                }}
            >
                <Sidebar />
                <div
                    className={'relative flex flex-col overflow-hidden w-full transition-all duration-300'}
                    style={{zIndex: 1}}
                >
                    <img
                        src={DashboardBgTop}
                        alt="Dashboard Top Bg Image"
                        className="absolute top-0 left-0 w-full z-[-1]"
                    />

                    <img
                        src={DashboardBgBottom}
                        alt="Dashboard Bottom Bg Image"
                        className="absolute bottom-0 left-0 w-full z-[-1]"
                    />

                    <div className="flex-1 overflow-y-auto relative z-10 p-4">
                        <Outlet />
                    </div>
                </div>
            </div>

            <Footer ref={footerRef} />
        </div>
    );
};
