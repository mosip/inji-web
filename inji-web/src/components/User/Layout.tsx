import React, {useEffect, useRef} from 'react';
import {Sidebar} from './Sidebar';
import {Header} from './Header';
import {Footer} from '../PageTemplate/Footer';
import {useSelector} from 'react-redux';
import {RootState} from '../../types/redux';
import {getCredentialTypeDisplayObjectForCurrentLanguage, getDirCurrentLanguage} from '../../utils/i18n';
import {Outlet} from 'react-router-dom';
import DashboardBgTop from '../../assets/Background.svg';
import DashboardBgBottom from '../../assets/DashboardBgBottom.svg';
import 'react-toastify/dist/ReactToastify.css';
import {CrossIconButton} from '../Common/Buttons/CrossIconButton';
import LayoutStyles from "../Common/LayoutStyles";
import {useDownloadSessionDetails} from '../../hooks/User/useDownloadSession';
import {useTranslation} from "react-i18next";
import {showToast} from "../Common/toast/ToastWrapper";
import {CloseButtonProps} from "react-toastify/dist/components";
import {RequestStatus} from "../../utils/constants";

export const Layout: React.FC = () => {
    const {t} = useTranslation('Layout')
    const language = useSelector((state: RootState) => state.common.language);
    const headerRef = useRef<HTMLDivElement>(null);
    const footerRef = useRef<HTMLDivElement>(null);
    const [headerHeight, setHeaderHeight] = React.useState(0);
    const [footerHeight, setFooterHeight] = React.useState(0);
    const {
        latestDownloadedSessionId,
        downloadInProgressSessions,
        setLatestDownloadedSessionId,
        currentSessionDownloadId,
        removeSession
    } = useDownloadSessionDetails();

    useEffect(() => {
        const updateHeights = () => {
            const headerHeight =
                headerRef.current?.getBoundingClientRect().height ?? 0;
            const footerHeight =
                footerRef.current?.getBoundingClientRect().height ?? 0;
            setHeaderHeight(headerHeight);
            setFooterHeight(footerHeight);
        };

        updateHeights();
        window.addEventListener('resize', updateHeights);
        return () => window.removeEventListener('resize', updateHeights);
    }, []);

    useEffect(() => {
            if (latestDownloadedSessionId && !currentSessionDownloadId) {
                const session = downloadInProgressSessions[latestDownloadedSessionId];
                if (session) {
                    const credentialTypeDisplayName =
                        getCredentialTypeDisplayObjectForCurrentLanguage(
                            session.credentialTypeDisplayObj,
                            language
                        ).name;
                    const {downloadStatus} = session;

                    const toastOptions = {
                        limit: 1,
                        autoClose: 3000,
                        icon: undefined,
                        closeButton: handleToasterCloseButton,
                        style: {
                            marginTop: headerHeight + 5,
                        },
                        className: ({type}: any) => {
                            const toastType = type ?? 'default';
                            return `${getToasterBackgroundColor(toastType)} ${LayoutStyles.toastContainerBase}`;
                        }
                    };
                    const toastMessage = downloadStatus === RequestStatus.DONE
                        ? t('VCDownload.success', {cardType: credentialTypeDisplayName})
                        : t('VCDownload.error', {cardType: credentialTypeDisplayName})

                    showToast({
                            message: toastMessage,
                            type: downloadStatus === RequestStatus.DONE ? 'success' : 'error',
                            options: {...toastOptions}
                        }
                    );
                    setLatestDownloadedSessionId(null);
                    removeSession(latestDownloadedSessionId);
                }
            }
        // eslint-disable-next-line react-hooks/exhaustive-deps
        }, [latestDownloadedSessionId, downloadInProgressSessions, setLatestDownloadedSessionId, removeSession]
    )
    ;

    const contextClassMap = {
        success: LayoutStyles.successToastContainer,
        error: LayoutStyles.errorToastContainer,
        default: ''
    };

    const getToasterBackgroundColor = (type: string) => {
        return contextClassMap[type as keyof typeof contextClassMap];
    };

    const handleToasterCloseButton = ({closeToast}: CloseButtonProps) => (
        <CrossIconButton
            onClick={(e) => {
                e.stopPropagation();
                closeToast(e);
            }}
            iconClassName={LayoutStyles.closeIcon}
        />
    );

    return (
        <div
            className="h-screen flex flex-col bg-iw-background font-base overflow-hidden w-full relative"
            dir={getDirCurrentLanguage(language)}
        >
            <Header headerRef={headerRef} headerHeight={headerHeight}/>

            <div
                className="flex flex-1 overflow-hidden w-full relative"
                style={{
                    marginTop: headerHeight,
                    marginBottom: footerHeight,
                    zIndex: 0
                }}
            >
                <Sidebar/>
                <div
                    className={
                        'relative flex flex-col overflow-hidden w-full transition-all duration-300'
                    }
                    style={{zIndex: 1}}
                >
                    <img
                        src={DashboardBgTop}
                        alt="Gradient Top Background"
                        className="absolute top-0 left-0 w-full z-[-1]"
                    />

                    <img
                        src={DashboardBgBottom}
                        alt="Gardient Bottom Background"
                        className="absolute bottom-0 left-0 w-full z-[-1]"
                    />

                    <div className="flex flex-grow flex-col overflow-y-auto relative z-10 p-4">
                        <Outlet />
                    </div>

                </div>
            </div>

            <Footer footerRef={footerRef}/>
        </div>
    );
};