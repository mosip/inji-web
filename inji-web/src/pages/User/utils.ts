import { User } from "../../components/Dashboard/types";
import { KEYS } from "../../utils/constants";
import {ROUTES} from "../../constants/Routes";

export const convertStringIntoPascalCase = (text: string | undefined) => {
    return (
        text?.toLocaleLowerCase()
            .split(' ')
            .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
            .join(' ')
    );
};

export const navigateToUserHome = (navigate: any) =>
    navigate(ROUTES.USER_HOME);

export const validateWalletUnlockStatus = (
    cachedWalletId: string | null,
    storageWalletId: string | null,
    navigate: (path: string) => void,
    user: User
) => {
    if (cachedWalletId && (cachedWalletId === storageWalletId)) {
        console.info('Wallet is unlocked!');
    } else {
        console.warn(
            'Wallet exists but is locked, redirecting to `/pin` to unlock the wallet.'
        );
        if(user) {
            navigate('/pin');
            localStorage.removeItem(KEYS.WALLET_ID);
        }
    }
};