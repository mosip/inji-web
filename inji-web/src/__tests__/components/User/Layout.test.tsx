import {setMockUseSelectorState } from '../../../test-utils/mockUtils';
import { render, screen, act } from '@testing-library/react';
import { Layout } from '../../../components/User/Layout';
import { useSelector } from 'react-redux';
import * as i18n from '../../../utils/i18n';

jest.mock('../../../components/User/Header', () => ({
  Header: ({ headerRef, headerHeight }: any) => (
    <div data-testid="Header" ref={headerRef}>
      Header - height: {headerHeight}
    </div>
  ),
}));

jest.mock('../../../components/User/Sidebar', () => ({
  Sidebar: () => <div data-testid="Sidebar">Sidebar</div>,
}));

jest.mock('../../../components/PageTemplate/Footer', () => ({
  Footer: ({ footerRef }: any) => (
    <div data-testid="Footer" ref={footerRef}>
      Footer
    </div>
  ),
}));

jest.mock('../../../utils/i18n', () => ({
  getDirCurrentLanguage: jest.fn(),
}));

jest.mock('../../../assets/Background.svg', () => 'mock-dashboard-bg-top');
jest.mock('../../../assets/DashboardBgBottom.svg', () => 'mock-dashboard-bg-bottom');

describe('Layout component', () => {
  beforeEach(() => {
    setMockUseSelectorState({ common: { language: 'en' } });
    (i18n.getDirCurrentLanguage as jest.Mock).mockReturnValue('ltr');
  });

  it('renders layout with Header, Sidebar, Outlet and Footer', () => {
    render(<Layout />);
    expect(screen.getByTestId('Header')).toBeInTheDocument();
    expect(screen.getByTestId('Sidebar')).toBeInTheDocument();
    expect(screen.getByTestId('Footer')).toBeInTheDocument();
    expect(screen.getByTestId('Outlet')).toBeInTheDocument();
  });

  it('sets dir attribute based on current language', () => {
    (i18n.getDirCurrentLanguage as jest.Mock).mockReturnValue('rtl');
    const { container } = render(<Layout />);
    expect(container.firstChild).toHaveAttribute('dir', 'rtl');
  });

  it('updates header and footer height on mount and resize', () => {
    const { container } = render(<Layout />);

    const header = screen.getByTestId('Header');
    const footer = screen.getByTestId('Footer');

    // Simulate DOM rects
    jest.spyOn(header, 'getBoundingClientRect').mockReturnValue({
      height: 80,
      width: 0, top: 0, left: 0, bottom: 0, right: 0, x: 0, y: 0, toJSON: () => {},
    });

    jest.spyOn(footer, 'getBoundingClientRect').mockReturnValue({
      height: 60,
      width: 0, top: 0, left: 0, bottom: 0, right: 0, x: 0, y: 0, toJSON: () => {},
    });

    act(() => {
      window.dispatchEvent(new Event('resize'));
    });

    expect(header).toBeInTheDocument();
    expect(footer).toBeInTheDocument();
  });

   // Snapshot test for the default (ltr) language rendering of Layout
   it('matches snapshot for default language', () => {
    const { container } = render(<Layout />);
    expect(container.firstChild).toMatchSnapshot();
  });

  // Snapshot test for RTL language rendering of Layout
  it('matches snapshot for rtl language', () => {
    (i18n.getDirCurrentLanguage as jest.Mock).mockReturnValue('rtl');
    const { container } = render(<Layout />);
    expect(container.firstChild).toMatchSnapshot();
  });

  // Snapshot test after simulating a resize that updates header and footer heights
  it('matches snapshot after header and footer heights update on resize', () => {
    const { container } = render(<Layout />);

    const header = screen.getByTestId('Header');
    const footer = screen.getByTestId('Footer');

    // Set DOM rects for header and footer to specific heights
    jest.spyOn(header, 'getBoundingClientRect').mockReturnValue({
      height: 80,
      width: 0,
      top: 0,
      left: 0,
      bottom: 0,
      right: 0,
      x: 0,
      y: 0,
      toJSON: () => {},
    });
    jest.spyOn(footer, 'getBoundingClientRect').mockReturnValue({
      height: 60,
      width: 0,
      top: 0,
      left: 0,
      bottom: 0,
      right: 0,
      x: 0,
      y: 0,
      toJSON: () => {},
    });

    act(() => {
      window.dispatchEvent(new Event('resize'));
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
