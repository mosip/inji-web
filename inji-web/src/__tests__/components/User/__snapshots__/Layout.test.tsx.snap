// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout component matches snapshot after header and footer heights update on resize 1`] = `
<div
  class="h-screen flex flex-col bg-iw-background font-base overflow-hidden w-full relative"
  dir="ltr"
>
  <div
    data-testid="Header"
  >
    Header - height: 
    80
  </div>
  <div
    class="flex flex-1 overflow-hidden w-full relative"
    style="margin-top: 80px; margin-bottom: 60px; z-index: 0;"
  >
    <div
      data-testid="Sidebar"
    >
      Sidebar
    </div>
    <div
      class="relative flex flex-col overflow-hidden w-full transition-all duration-300"
      style="z-index: 1;"
    >
      <img
        alt="Gradient Top Background"
        class="absolute top-0 left-0 w-full z-[-1]"
        src="mock-dashboard-bg-top"
      />
      <img
        alt="Gardient Bottom Background"
        class="absolute bottom-0 left-0 w-full z-[-1]"
        src="mock-dashboard-bg-bottom"
      />
      <div
        class="flex flex-grow flex-col relative z-10 overflow-y-auto"
      >
        <div
          class="flex flex-col items-center justify-center min-h-full w-full"
        >
          <div
            data-testid="Outlet"
          >
            Page content
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="Footer"
  >
    Footer
  </div>
</div>
`;

exports[`Layout component matches snapshot for default language 1`] = `
<div
  class="h-screen flex flex-col bg-iw-background font-base overflow-hidden w-full relative"
  dir="ltr"
>
  <div
    data-testid="Header"
  >
    Header - height: 
    0
  </div>
  <div
    class="flex flex-1 overflow-hidden w-full relative"
    style="margin-top: 0px; margin-bottom: 0px; z-index: 0;"
  >
    <div
      data-testid="Sidebar"
    >
      Sidebar
    </div>
    <div
      class="relative flex flex-col overflow-hidden w-full transition-all duration-300"
      style="z-index: 1;"
    >
      <img
        alt="Gradient Top Background"
        class="absolute top-0 left-0 w-full z-[-1]"
        src="mock-dashboard-bg-top"
      />
      <img
        alt="Gardient Bottom Background"
        class="absolute bottom-0 left-0 w-full z-[-1]"
        src="mock-dashboard-bg-bottom"
      />
      <div
        class="flex flex-grow flex-col relative z-10 overflow-y-auto"
      >
        <div
          class="flex flex-col items-center justify-center min-h-full w-full"
        >
          <div
            data-testid="Outlet"
          >
            Page content
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="Footer"
  >
    Footer
  </div>
</div>
`;

exports[`Layout component matches snapshot for rtl language 1`] = `
<div
  class="h-screen flex flex-col bg-iw-background font-base overflow-hidden w-full relative"
  dir="rtl"
>
  <div
    data-testid="Header"
  >
    Header - height: 
    0
  </div>
  <div
    class="flex flex-1 overflow-hidden w-full relative"
    style="margin-top: 0px; margin-bottom: 0px; z-index: 0;"
  >
    <div
      data-testid="Sidebar"
    >
      Sidebar
    </div>
    <div
      class="relative flex flex-col overflow-hidden w-full transition-all duration-300"
      style="z-index: 1;"
    >
      <img
        alt="Gradient Top Background"
        class="absolute top-0 left-0 w-full z-[-1]"
        src="mock-dashboard-bg-top"
      />
      <img
        alt="Gardient Bottom Background"
        class="absolute bottom-0 left-0 w-full z-[-1]"
        src="mock-dashboard-bg-bottom"
      />
      <div
        class="flex flex-grow flex-col relative z-10 overflow-y-auto"
      >
        <div
          class="flex flex-col items-center justify-center min-h-full w-full"
        >
          <div
            data-testid="Outlet"
          >
            Page content
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="Footer"
  >
    Footer
  </div>
</div>
`;
